File: lib/models/customer.dart
class Customer {
  final String id;
  final String businessId;
  final String name;
  final String? gender; // 可選字段
  final String? phone; // 可選字段
  final String? email; // 可選字段
  final bool isArchived;
  final bool needsMerge;
  final bool isSpecialCustomer;
  final String? source;
  final DateTime createdAt;
  final DateTime updatedAt;

  Customer({
    required this.id,
    required this.businessId,
    required this.name,
    this.gender,
    this.phone,
    this.email,
    this.isArchived = false,
    this.needsMerge = false,
    this.isSpecialCustomer = false,
    this.source,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Customer.fromJson(Map<String, dynamic> json) {
    return Customer(
      id: json['id'],
      businessId: json['business_id'],
      name: json['name'],
      gender: json['gender'],
      phone: json['phone'],
      email: json['email'],
      isArchived: json['is_archived'] ?? false,
      needsMerge: json['needs_merge'] ?? false,
      isSpecialCustomer: json['is_special_customer'] ?? false,
      source: json['source'],
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at']) 
          : DateTime.now(),
      updatedAt: json['updated_at'] != null 
          ? DateTime.parse(json['updated_at']) 
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'business_id': businessId,
      'name': name,
      'gender': gender,
      'phone': phone,
      'email': email,
      'is_archived': isArchived,
      'needs_merge': needsMerge,
      'is_special_customer': isSpecialCustomer,
      'source': source,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
} 

File: backend/internal/models/customer.go
package models

import (
	"time"

	"gorm.io/gorm"
)

// Customer represents the customer model
type Customer struct {
	ID                string         `gorm:"primaryKey;type:varchar(36)" json:"id"`
	BusinessID        string         `gorm:"type:varchar(36);not null;index" json:"business_id"`
	Name              string         `gorm:"type:varchar(100);not null" json:"name"`
	Gender            *string        `gorm:"type:varchar(10)" json:"gender,omitempty"`
	Phone             *string        `gorm:"type:varchar(20);index" json:"phone,omitempty"`
	Email             *string        `gorm:"type:varchar(255);index" json:"email,omitempty"`
	IsArchived        bool           `gorm:"default:false" json:"is_archived"`
	NeedsMerge        bool           `gorm:"default:false" json:"needs_merge"`
	IsSpecialCustomer bool           `gorm:"default:false" json:"is_special_customer"`
	Source            *string        `gorm:"type:varchar(50)" json:"source,omitempty"`
	CreatedAt         time.Time      `json:"created_at"`
	UpdatedAt         time.Time      `json:"updated_at"`
	DeletedAt         gorm.DeletedAt `gorm:"index" json:"-"`

	// Relationships
	Business *Business `gorm:"foreignKey:BusinessID" json:"business,omitempty"`
}

// TableName specifies the table name for Customer model
func (Customer) TableName() string {
	return "customers"
}

// BeforeCreate hook to generate UUID for customer ID
func (c *Customer) BeforeCreate(tx *gorm.DB) error {
	if c.ID == "" {
		c.ID = generateUUID()
	}
	return nil
}

// CreateCustomerRequest represents the create customer request payload
type CreateCustomerRequest struct {
	BusinessID        string `json:"business_id" binding:"required"`
	Name              string `json:"name" binding:"required,min=2,max=100"`
	Gender            string `json:"gender,omitempty"`
	Phone             string `json:"phone,omitempty"`
	Email             string `json:"email,omitempty"`
	IsArchived        bool   `json:"is_archived,omitempty"`
	NeedsMerge        bool   `json:"needs_merge,omitempty"`
	IsSpecialCustomer bool   `json:"is_special_customer,omitempty"`
	Source            string `json:"source,omitempty"`
}

// UpdateCustomerRequest represents the update customer request payload
type UpdateCustomerRequest struct {
	Name              string `json:"name" binding:"omitempty,min=2,max=100"`
	Gender            string `json:"gender,omitempty"`
	Phone             string `json:"phone,omitempty"`
	Email             string `json:"email,omitempty"`
	IsArchived        *bool  `json:"is_archived,omitempty"`
	NeedsMerge        *bool  `json:"needs_merge,omitempty"`
	IsSpecialCustomer *bool  `json:"is_special_customer,omitempty"`
	Source            string `json:"source,omitempty"`
}

// ToModel converts CreateCustomerRequest to Customer model
func (r *CreateCustomerRequest) ToModel() *Customer {
	customer := &Customer{
		BusinessID:        r.BusinessID,
		Name:              r.Name,
		IsArchived:        r.IsArchived,
		NeedsMerge:        r.NeedsMerge,
		IsSpecialCustomer: r.IsSpecialCustomer,
	}

	if r.Gender != "" {
		customer.Gender = &r.Gender
	}
	if r.Phone != "" {
		customer.Phone = &r.Phone
	}
	if r.Email != "" {
		customer.Email = &r.Email
	}
	if r.Source != "" {
		customer.Source = &r.Source
	}

	return customer
}

// IsActive returns true if the customer is not archived
func (c *Customer) IsActive() bool {
	return !c.IsArchived
}
